Introduce executor for concurrent search

javanna opened the pull request:
## Enable concurrent search across segments for knn queries

Elasticsearch has historically performed search sequentially across the segments. Lucene supports parallelizing search across segments when collecting hits (via collector managers) as well as when rewriting certain queries (e.g. knn queries).

This commit also enables concurrent search execution in the DFS phase, which is going to improve resource usage as well as performance of knn queries which benefit from both concurrent rewrite and collection.

## Enable offloading of sequential collection to search worker thread pool

We will enable concurrent execution for the query phase in a subsequent commit. While this commit does not introduce parallelism for the query phase, it introduces offloading sequential computation to the newly introduced executor. This is true both for situations where a single slice needs to be searched, as well as scenarios where a specific request does not support concurrency (currently only DFS phase does regardless of the request). Sequential collection is not offloaded only if the request includes aggregations that don't support offloading: composite, nested and cardinality as their post collection method must be executed in the same thread as the collection or we'll trip a lucene assertion that verifies that doc_values are pulled and consumed from the same thread.

## Technical details

Elasticsearch is now ready to support concurrency within a single shard. Search is already performed using collector managers, and the last missing piece is providing an executor to the index searcher so that it can offload the concurrent computation to it.

This commit introduces a secondary executor, used exclusively to execute the concurrent bits of search. The search threads are still the ones that coordinate the search (where the caller search will originate from), but the actual work will be offloaded to the newly introduced executor.

We are offloading not only parallel execution but also sequential execution, to make the workload more predictable, as it would be surprising to have bits of search executed in either of the two thread pools. Also, that would introduce the possibility to suddenly run a higher amount of heavy operations overall (some in the caller thread and some in the separate threads), which could overload the system as well as make sizing of thread pools more difficult.

Note that fetch, together with other actions,  is still executed in the search thread pool. This commit does not make the search thread pool merely a coordinating only thread pool, It does so only for what concerns the IndexSearcher#search operation itself, which is though a big portion of the different phases of search API execution.

Given that the searcher blocks waiting for all tasks to be completed, we take a simple approach of introducing a thread pool executor that has the same size as the existing search thread pool but relies on an unbounded queue. This simplifies handling of thread pool queue and rejections. In fact, we'd like to guarantee that the secondary thread pool won't reject, and delegate queuing entirely to the search thread pool which is the entry point for every search operation anyway. The principle behind this is that if you got a slot in the search thread pool, you should be able to complete your search, and rather quickly.

As part of this commit we are also introducing the ability to cancel tasks that have not started yet, so that if any task throws an exception, other tasks are prevented from starting needless computation.

Relates to #80693
Relates to #90700

javanna committed:
Introduce executor for concurrent search

Elasticsearch has historically performed search sequentially across the segments.
Lucene supports parallelizing search across segments when collecting hits (via collector managers)
as well as when rewriting certain queries (e.g. knn queries).

Elasticsearch is now ready to support concurrency within a single shard
too. Search is already performed using collector managers, and the last
piece that is missing is providing an executor to the index searcher
so that it can offload the concurrent computation to it.

This commit introduces a secondary executor, used exclusively to execute
the concurrent bits of search. The search threads are still the ones that
coordinate the search (where the caller search will originate from), but
the actual work will be offloaded to the newly introduced executor.

We are offloading not only parallel execution but also sequential
execution, to make the workload more predictable, as it would be
surprising to have bits of search executed in either of the two thread pools.
Also, that would introduce the possibility to suddenly run a higher
amount of heavy operations overall (some in the caller thread and some
in the separate threads), which could overload the system as well as
make sizing of thread pools more difficult.

Note that fetch, together with other actions,  is still executed in the
search thread pool. This commit does not make the search thread pool merely
a coordinating only thread pool, It does so only for what concerns the
IndexSearcher#search operation itself, which is though a big portion of
the different phases of search API execution.

Given that the searcher blocks waiting for all tasks to be completed,
we take a simple approach of introducing a bounded executor, which
blocks whenever there are no threads to directly execute tasks. This simplifies
handling of thread pool queue and rejections. In fact, we can guarantee
that the secondary thread pool won't reject, and delegate queueing
entirely to the search thread pool which is the entry point for every
search operation anyway. The principle behind this is that if you got a
slot in the search thread pool, you should be able to complete your
search.

As part of this commit we are also introducing the ability to cancel
tasks that have not started yet, so that if any task throws an exception,
other tasks are prevented from starting needless computation.

This commit also enables concurrenct search execution in the DFS phase,
which is going to improve resource usage as well as performance of knn
queries which benefit from both concurrent rewrite and collection.

We will enable concurrent execution for the query phase in a subsequent
commit.

Relates to #80693
Relates to #90700

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @javanna, I've created a changelog YAML for you.

javanna committed:
Update docs/changelog/98204.yaml

javanna committed:
forbidden api

benwtrent commented on server/src/main/java/org/elasticsearch/common/util/concurrent/BoundedExecutor.java:
```diff
@@ -0,0 +1,76 @@
+        try {
+            semaphore.acquire();
+        } catch (InterruptedException e) {
+            throw new ThreadInterruptedException(e);
```
If the thread is interrupted, shouldn't we bubble it up via `Thread.currentThread().interrupt();`?

javanna committed:
grrrrrrrrrr

javanna committed:
leftover

benwtrent commented on server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java:
```diff
@@ -141,25 +145,27 @@ private ContextIndexSearcher(
+                this.leafSlices = new LeafSlice[]{new LeafSlice(new ArrayList<>(getLeafContexts()))};
+            } else {
+                this.leafSlices = computeSlices(getLeafContexts(), executor.getBound(), minimumDocsPerSlice);
+                assert (this.leafSlices.length <= executor.getBound());
```
could you add an assertion message here? I always find those helpful when I get an angry assertion stacktrace during testing :)

javanna committed:
forbidden api

benwtrent commented on server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:
```diff
@@ -181,220 +187,91 @@ public void collect(int doc, long bucket) throws IOException {
+                KnnFloatVectorQuery vectorQuery = new KnnFloatVectorQuery("float_vector", new float[] { 0, 0, 0 }, 10, null);
+                Query rewrittenQuery = vectorQuery.rewrite(searcher);
+                // Note: we expect one execute calls less than segments since the last is executed on the caller thread.
+                // For details see QueueSizeBasedExecutor#processTask
```
```suggestion
                // For details see BoundedExecutor#processTask
```

javanna replied:
this is actually correct, it refers to the executor currently used in Lucene.

benwtrent replied:
😕 my bad.

javanna committed:
spotless

benwtrent commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1069,7 +1071,8 @@ private DefaultSearchContext createSearchContext(
                 fetchPhase,
                 lowLevelCancellation,
-                this.enableConcurrentCollection && concurrentSearchEnabled(resultsType, request.source())
+                this.enableConcurrentCollection ? concurrentSearchExecutor : null,
```
It seems weird to me that we now have two settings, `enableConcurrentCollection` set by `SEARCH_CONCURRENCY_ENABLED` and `SEARCH_CONCURRENT.size == 1`. 

Is it worth it to set `SEARCH_CONCURRENCY_ENABLED` when a user can set ``SEARCH_CONCURRENT.size = 1`?

javanna replied:
There is a subtle difference: one entirely disables concurrency, meaning the concurrent thread pool is no longer used. That is an escape hatch that is currently undocumented.

The force single slice one instead depends on the phase and the request, and forces the generation of a single slice, which though will be offloaded to the concurrent thread pool. 

This distinction is important internally to make sure that once we execute concurrently either one of the thread pools does the heavy work, otherwise both may do it and that could overload the system and make sizing more difficult.

javanna replied:
I simplified this: we were already providing a max num slices, I can just set that to 1 instead of requiring an additional forceSingleSlice argument.

benwtrent reviewed (commented):
I am by no means an expert here and my review doesn't count for much, I had some minor things

henningandersen reviewed (commented):
Started on this and left a few initial comments.

javanna commented on server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java:
```diff
@@ -141,25 +145,27 @@ private ContextIndexSearcher(
+        } else {
+            if (forceSequentialCollection) {
+                // we offload to the executor unconditionally, including requests that don't support concurrency
+                // TODO are we good offloading aggs that don't support concurrency to the concurrent executor or is postCollect an issue?
```
yes that aspect is covered but postCollect at times does the entire collection and that needs to happen on the same thread where the collection itself happens. I lost track of where we are with that, and whether offloading may cause problems or not. Hopefully we'll see test failures if so, as this PR does not enable concurrency for the query phase but it does enable the offloading to the separate thread pool for sequential execution.

javanna replied:
To share more details, the problem used to be assertions in Lucene that verified that doc_values read as part of post collect were being read in the same thread where the collection happens. We said the aggs that require this may not support concurrency to start with, yet they may run into issues even when running them sequentially only because we offload them to the separate pool. To be verified.

javanna replied:
@martijnvg do you have thoughts on this? Is it ok to always offload the collection even if sequential (single slice) including for requests that include aggs that explicitly don't support concurrency?

javanna replied:
Martijn and I chatted and spotted three aggs that are problematic, for which we are disabling offloading sequential collection (nested, composite and cardinality). There have been IT tests failures around these in this PR, which is good. A good explanation of the different aggs that are problematic when it comes to inter-segment concurrency is at #96023 and it clearly explains the two category of problems, out of which one involves not only parallel collection but also sequential collection on a separate thread.

As a follow-up we will see whether we can move postCollect to the worker thread and come up with a long term plan for these.

javanna replied:
postCollect has been moved to the search workers, and this is no longer a problem.

javanna committed:
Update server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

Co-authored-by: Henning Andersen <33268011+henningandersen@users.noreply.github.com>

javanna committed:
iter

javanna commented on server/src/main/java/org/elasticsearch/common/util/concurrent/BoundedExecutor.java:
```diff
@@ -0,0 +1,76 @@
+ * That means that we'll accept tasks before there are free threads to take them, hence there will be some overflow queueing required in
+ * the underlying thread pool executor. This is the reason why a queue is required despite the wait applied at submit.
+ */
+public class BoundedExecutor implements Executor {
```
Do you mean using a SynchronousQueue as the queue for the worker thread pool, without introducing this BoundedExecutor, or in combination with the bounded executor? I am not entirely following yet how that would work. What would be pros and cons?

henningandersen replied:
Yes, the former. The main benefit is avoiding the extra code and the small hack to still have a queue for the duration between the task completing and the thread being able to pick up new work. Just seems intuitive now that it is supported by the jdk itself.

javanna replied:
Ok, I will look into this and report back my findings.

javanna replied:
I think being able to use the thread pool directly without wrapping is also an advantage. Let's see what that looks like.

javanna replied:
Heads up: it seems to work well, looks like a drop-in replacement in my current tests. I need to hash out how to adjust some of the details of the cancellation tests but this looks very promising. I will move forward with the synchronous queue approach.

javanna committed:
iter

javanna commented on server/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java:
```diff
@@ -502,7 +502,7 @@ public void apply(Settings value, Settings current, Settings previous) {
         SearchModule.INDICES_MAX_CLAUSE_COUNT_SETTING,
         SearchModule.INDICES_MAX_NESTED_DEPTH_SETTING,
-        SearchModule.SEARCH_CONCURRENCY_ENABLED,
+        SearchService.SEARCH_WORKER_THREADS_ENABLED,
```
I have renamed the escape hatch to disable concurrency to align it with the name of the new thread pool, as it effectively affects whether the new thread pool is enabled or not.

javanna commented on server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:
```diff
@@ -150,19 +153,28 @@ final class DefaultSearchContext extends SearchContext {
+                forceSequentialCollection,
+                minimumDocsPerSlice,
+                // TODO is there a better way that does not require casting the executor?
+                ((ThreadPoolExecutor) executor).getMaximumPoolSize()
```
this cast is not great. We had it before, and I have made it effective only when concurrency is enabled. There was SnapshotResiliencyTests that injects a different executor and was causing a class cast exception here, to I went for disabling concurrency in that test entirely via the cluster setting.

We could also expose the max pool size (searchOrGetThreadPoolSize) in some way in ThreadPool instead of retrieving it from the thread pool executor, but I am not too sure that that is better.

javanna commented on server/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -281,13 +281,6 @@ public class SearchModule {
-        Setting.Property.NodeScope,
-        Setting.Property.Dynamic
-    );
-
```
I moved this to SearchService together with the other existing search settings (including minimum docs per slice which affects search concurrency)

javanna commented:
@henningandersen @DaveCTurner I pushed an update to replace the bounded executor with a plain thread pool executor that relies on a synchronous queue. I added some questions for you two that I could use your help with. 

I am still working on the additional ContextIndexSearcherTests, that I just realized are not using the synchronous queue.  Need to update those, work in progress.

DaveCTurner reviewed (commented):
I don't think a `SynchronousQueue` has the right semantics here. AIUI it means that each `.execute()` call will wait for the task to start executing on a worker thread, but I think we want to block the thread calling `.execute()` until the task it submits is _complete_. Am I missing something there?

henningandersen commented:
> I don't think a `SynchronousQueue` has the right semantics here. AIUI it means that each `.execute()` call will wait for the task to start executing on a worker thread, but I think we want to block the thread calling `.execute()` until the task it submits is _complete_. Am I missing something there?

Blocking waiting for completion should be done by calling `Future.get()` on the result of `execute`. What `SynchronousQueue` helps with is to avoid any queuing and block `execute` calls until a free thread is available in the search worker pool.

DaveCTurner commented:
`.execute()` doesn't return anything, but yes to the more general point that we'll need to wrap each task with some kind of `Future` on which we wait after calling `execute()`.

> What `SynchronousQueue` helps with is to avoid any queuing and block `execute` calls until a free thread is available in the search worker pool.

Sure, but can you help me understand why we need that? Is it just performance or are there semantic differences I'm missing? If it's performance, how much time does it save in practice? I'd prefer not to introduce yet another kind of queue (and another source of potential blocking to trip up the unwary) without a strong justification.

benwtrent commented on server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java:
```diff
@@ -211,9 +211,7 @@ public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBui
-                // TODO the queue is only needed as overflow because we may release permits before the executor updated its internal state.
-                // Queue size equal to the bound (max pool size) should be enough to ensure no rejection, but it could also be unbounded.
-                searchOrGetThreadPoolSize,
+                -128, // TODO clean this up
```
We already use "special" numbers to indicate an action ('-1' is one such number). Using a `-128` seems, eh?

Why not add a constructor to `FixedExecutorBuilder` that accepts a manually created queue?

benwtrent commented on server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:
```diff
@@ -127,11 +128,24 @@ public static EsThreadPoolExecutor newFixed(
     ) {
         BlockingQueue<Runnable> queue;
-        if (queueCapacity < 0) {
+        if (queueCapacity == -128) {
```
We already use "special" numbers to indicate an action ('-1' is one such number). Using a -128 seems, eh?

Why not add a constructor to FixedExecutorBuilder that accepts a manually created queue? Are we afraid of callers doing something foolish?

javanna commented:
> we'll need to wrap each task with some kind of Future on which we wait after calling execute().

This is already the case, it's how the `FutureTask` are created in `ContextIndexSearcher`, and `Future#get` is called on them.

javanna committed:
iter

javanna committed:
Merge branch 'main' into enhancement/bounded_executor_concurrent_search

javanna committed:
iter

javanna commented on server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java:
```diff
@@ -150,19 +153,27 @@ final class DefaultSearchContext extends SearchContext {
+                executor,
+                minimumDocsPerSlice,
+                // TODO is there a better way that does not require casting the executor?
+                forceSequentialCollection ? 1 : ((ThreadPoolExecutor) executor).getMaximumPoolSize()
```
this cast is not great. We had it before, and I have made it effective only when concurrency is enabled. There was SnapshotResiliencyTests that injects a different executor and was causing a class cast exception here, to I went for disabling concurrency in that test entirely via the cluster setting.

We could also expose the max pool size (searchOrGetThreadPoolSize) in some way in ThreadPool instead of retrieving it from the thread pool executor, but I am not too sure that that is better.

DaveCTurner replied:
I think it might be a bit nicer to handle this case locally rather than changing settings in `SnapshotResiliencyTests` (which might never get reverted) - something like `executor instanceof ThreadPoolExecutor threadPoolExecutor ? threadPoolExecutor.getMaximumPoolSize() : 1` perhaps?

javanna replied:
Ok, I am not sure which is best: disable concurrency in the test that has the funky thread pool or silently disable it in prod code whenever you have a funky thread pool. The latter seems scary to me, is it a situation that could happen outside of tests?

DaveCTurner replied:
If you're using an executor that isn't from a threadpool then its max pool size is effectively `1` anyway (it's a thread on its own). I think I'd rather fall back to some sensible behaviour instead of failing with a `ClassCastException` if ever we encounter an inappropriate executor, but I'm also pretty sure we wouldn't see an inappropriate executor in prod code today.

Perhaps slightly neater would be to introduce a new interface with a method for this specific case, and apply it to both `EsThreadPoolExecutor` and the ones used in tests. I can see that being more generally useful, there are several other places where we try and size some pool of workers according to the size of the underlying threadpool.

javanna replied:
I went for your initial suggestion, thanks for the feedback.

javanna committed:
iter

javanna commented on server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java:
```diff
@@ -651,34 +518,365 @@ public boolean isCacheable(LeafReaderContext ctx) {
+     * Simulate having N threads busy doing other work (e.g. other searches) otherwise all slices can be executed directly, given that
+     * the number of slices is dependent on the max pool size.
+     */
+    public void testCancelSliceTasksOnException() throws Exception {
```
I have adjusted and simplified some existing tests, as well as added two tests are cancellation of tasks on exception / timeout. I wonder if there's more tests that we'd like to add to verify the behaviour of the new thread pool in combination with context index searcher.

javanna committed:
special case for no segments

javanna committed:
spotless

javanna committed:
terminate thread pool in test

javanna committed:
Merge branch 'main' into enhancement/bounded_executor_concurrent_search

javanna committed:
fix thread pool executor cast

javanna committed:
iter

javanna committed:
Merge branch 'main' into enhancement/bounded_executor_concurrent_search

javanna committed:
iter

javanna commented on server/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java:
```diff
@@ -334,15 +334,34 @@ public boolean isInSortOrderExecutionRequired() {
          */
-        public boolean supportsConcurrentExecution() {
+        // TODO this is not used yet, it will be used once we enable parallel collection for the query phase
+        public boolean supportsParallelCollection() {
```
I renamed this to hopefully be more accurate about what it controls, especially as I introduced the new supportsOffloadingSequentialCollection below.

javanna committed:
iter

javanna committed:
remove TODO and clarify comments

javanna committed:
clarify docs

benwtrent commented on server/src/main/java/org/elasticsearch/search/aggregations/AggregatorFactories.java:
```diff
@@ -334,15 +334,34 @@ public boolean isInSortOrderExecutionRequired() {
                 }
             }
-            return isInSortOrderExecutionRequired() == false;
+            return true;
```
Should this method also return `supportsOffloadingSequentialCollection()` here instead of `true`? Seems like we should be careful here as now we have two things controlling parallelism. Don't want them getting out of sync.

javanna replied:
I spent the best part of my day trying to make this work. I don't like the need for two knobs here and it's not straight-forward what they do. My hope is that we can get rid of both soon.

These two methods are really just exposing the value returned by the aggregation builders, and that already includes the fallback logic, so we should be fine. I will though add additional unit tests for this, and verify that the two are aligned.

javanna committed:
spotless

benwtrent reviewed (approved):
One comment about the aggregators

javanna committed:
iter

jdconrad reviewed (approved):
LGTM!

I don't see any caveats here related to executing multiple queries for ranking. As it stands each query will be executed sequentially and once concurrency is added as part of the query phase each one will benefit if it's able. The only consideration is if we want to parallelize all the queries to run at the same time which could have significant benefit if ranking becomes a heavily used feature.

One question -- is the behavior of the search workers thread pool such that it will still behave as a FIFO for queries coming in where all the threads you'd want to launch for a search are queued right at the start? Or is there the possibility of thrashing where other searches can cut in line? I hope this question makes sense.

DaveCTurner commented on server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java:
```diff
@@ -333,13 +333,21 @@ public <C extends Collector, T> T search(Query query, CollectorManager<C, T> col
      * */
     private <C extends Collector, T> T search(Weight weight, CollectorManager<C, T> collectorManager, C firstCollector) throws IOException {
-        if (queueSizeBasedExecutor == null || leafSlices.length <= 1) {
+        // the executor will be null only when concurrency is disabled at the cluster level
```
This comment isn't true and that's a little confusing, we also have no executor if the search doesn't support it.

javanna replied:
yea the comment was outdated. But is no longer true after the latest updates. There are no exceptions to offloading, so the comment is true again, and we are good :)

martijnvg committed:
Remove supportsOffloadingSequentialCollection() method and do aggregation postcollection() in ContextIndexSearcher on the search worker thread.

The way how postcollection gets invoked is dirty because in the ContextIndexSearcher only knows about lucene collectors and not BucketCollector. So a few casts are involved to make this work.

martijnvg committed:
Also postCollect when there are no slices

martijnvg committed:
address tests

martijnvg committed:
spotless

martijnvg committed:
Correctly overwrite supportsParallelCollection(...)

martijnvg committed:
fixed unit tests

martijnvg committed:
Also perform postCollection when executing search in sort order.

javanna commented on server/src/main/java/org/elasticsearch/search/aggregations/metrics/CardinalityAggregationBuilder.java:
```diff
@@ -113,7 +113,7 @@ public boolean supportsSampling() {
 
     @Override
-    public boolean supportsOffloadingSequentialCollection() {
+    public boolean supportsParallelCollection() {
```
If these could not support concurrency due to the post collect issue, does it mean that they can now support concurrency perhaps?

martijnvg replied:
For `cardinality`, I think yes. But let's change this when parallel collection is enabled in the query phase?

javanna replied:
Yes

javanna commented:
@jdconrad 

> One question -- is the behavior of the search workers thread pool such that it will still behave as a FIFO for queries coming in where all the threads you'd want to launch for a search are queued right at the start? Or is there the possibility of thrashing where other searches can cut in line? I hope this question makes sense.

There can be interleaving for requests that execute their subtasks at the same time, but the execution is indeed FIFO based.

javanna commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1060,16 +1064,21 @@ private DefaultSearchContext createSearchContext(
+            ExecutorService executor = this.enableSearchWorkerThreads
+                && supportsOffloadingSequentialCollection(resultsType, request.source()) ? threadPool.executor(Names.SEARCH_WORKER) : null;
+            int maximumNumberOfSlices = executor instanceof ThreadPoolExecutor tpe
+                && supportsParallelCollection(resultsType, request.source()) ? tpe.getMaximumPoolSize() : 1;
```
Update:

javanna commented on server/src/main/java/org/elasticsearch/search/aggregations/AggregationPhase.java:
```diff
@@ -72,6 +72,7 @@ private static void executeInSortOrder(SearchContext context, BucketCollector co
         searcher.setProfiler(context);
         try {
             searcher.search(context.rewrittenQuery(), collector);
+            collector.postCollection();
```
I am thinking that this is another exception to the offloading that we should remove. We should provide an executor to the time series index searcher, and offload its sequential execution to it.

martijnvg replied:
I was thinking the same thing, when making that change.
However this isn't a normal searcher, but a `TimeSeriesIndexSearcher` and does searching completely different.

javanna replied:
yep, I think though that because it's part of the query phase, it could potentially cause us to take in more heavy work than desired, say that we have many searches being offloaded and at the same time many others using this time series index searcher, we'd have many threads (potentially double compared to before adding the search workers thread pool) doing I/O and CPU bound stuff.

I don't think this is a blocker for this PR but we should look into it as a follow-up.

javanna committed:
Comment

elasticsearchmachine commented:
@javanna according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is labelled `release highlight` but the changelog has no `highlight` section

elasticsearchmachine commented:
@javanna according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is labelled `release highlight` but the changelog has no `highlight` section
